### PR TITLE
Implement communication component

### PR DESCRIPTION
**Pull request by al1img**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "cSpell.words": [
+        "cppcheck",
+        "Incrementer",
+        "zassert"
+    ]
+}
```
Comment (MykolaSolyanko) [inline .vscode/settings.json:7]: i think this files should be deleted

File: .vscode/c_cpp_properties.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+            "defines": [],
+            "compilerPath": "/usr/bin/gcc",
+            "cStandard": "gnu17",
+            "cppStandard": "c++14",
+            "intelliSenseMode": "linux-gcc-x64",
+            "configurationProvider": "ms-vscode.cmake-tools"
+        }
+    ],
+    "version": 4
+}
```
Comment (MykolaSolyanko) [inline .vscode/c_cpp_properties.json:17]: i think in this commit should be removed some files which related to vscode and test

File: src/clocksync/clocksync.hpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    /**
+     * Synchronizes system clock.
+     *
+     * @param time current time to set.
+     * @return aos::Error
+     */
+    virtual aos::Error ClockSync(const aos::time::Time& time) = 0;
+};
+
+#endif
```
Comment (MykolaSolyanko) [inline src/clocksync/clocksync.hpp:67]: I think these files are not relevant to this commit, I'm right?

File: src/domains/domd/domain_bins.S [hunk 4]
```diff
@@ -20,3 +20,4 @@ __dtb_ipl_start:
 .align 8
 .incbin CONFIG_DOMD_DTB_PATH
 __dtb_ipl_end:
+
```
Comment (MykolaSolyanko) [inline src/domains/domd/domain_bins.S:23]: Is this change necessary?

File: src/resourcemanager/resourcemanager.hpp [hunk 5]
```diff
@@ -11,19 +11,57 @@
 #include <aos/common/tools/error.hpp>
 #include <aos/common/tools/noncopyable.hpp>
 
+/**
+ * Resource manager interface.
+ */
+
```
Comment (mykola-kobets-epam) [inline src/resourcemanager/resourcemanager.hpp:17]: It seem to be rather ConfigurationManager

File: tests/communication/CMakeLists.txt [hunk 6]
```diff
@@ -52,4 +52,7 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
 # Target
 # ######################################################################################################################
 
-target_sources(app PRIVATE ../../src/communication/communication.cpp src/main.cpp)
```
Comment (mykola-kobets-epam) [inline tests/communication/CMakeLists.txt]: I think it would be better to link library under testing instead of adding source files.

File: src/utils/pbconvert.hpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+ * @param dst Aos string to convert to.
+ */
+template <size_t cSize>
+constexpr void PBToString(const char (&src)[cSize], aos::String& dst)
+{
+    auto len = strlen(src);
+
+    if (len > cSize - 1) {
+        len = cSize - 1;
+    }
```
Comment (mykola-kobets-epam) [inline src/utils/pbconvert.hpp:40]: I think it would be better to use assert here.

File: src/utils/pbconvert.hpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+template <typename T, size_t cSize>
+constexpr void PBToEnum(const char (&src)[cSize], T& dst)
+{
+    auto len = strlen(src);
+
+    if (len > cSize - 1) {
+        len = cSize - 1;
+    }
+
+    dst.FromString(aos::String(src, len));
```
Comment (mykola-kobets-epam) [inline src/utils/pbconvert.hpp:62]: Taking into account FromString could fail either assert check should be used or function should return error.

File: src/utils/pbconvert.hpp [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+        return;
+    }
+
+    str.Convert(aosError);
+}
+
+/**
+ * Converts Aos byte array to PB byte array.
+ *
+ * @tparam T PB byte array type.
```
Comment (mykola-kobets-epam) [inline src/utils/pbconvert.hpp:88]: It's strange that this function uses general type for a PB byte array.

File: src/communication/cmclient.cpp [hunk 10]
```diff
@@ -74,6 +78,80 @@ aos::Error CMClient::ProcessMessage(const aos::String& methodName, uint64_t requ
...
+
+    auto err = mResourceManager->CheckUnitConfig(pbUnitConfig.vendor_version, pbUnitConfig.unit_config);
+    if (!err.IsNone()) {
+        ErrorToPB(err, pbConfigStatus.error);
+    }
+
+    aos::StaticString<aos::cVendorVersionLen> version;
+
+    PBToString(pbUnitConfig.vendor_version, version);
+    StringToPB(version, pbConfigStatus.vendor_version);
```
Comment (mykola-kobets-epam) [inline src/communication/cmclient.cpp:292]: I think it should be possible to copy pbUnitConfig.vendor_version into pbConfigStatus.vendor_version without additional manipulations.

File: src/communication/vchannel.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+aos::Error VChannel::Read(aos::Array<uint8_t>& data, size_t size)
+{
+    auto err = data.Resize(size);
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    auto ret = vch_read(&mReadHandle, data.Get(), size);
+    if (ret < 0) {
+        return AOS_ERROR_WRAP(ret);
```
Comment (mykola-kobets-epam) [inline src/communication/vchannel.cpp:56]: I dont think there is a proper convertion from vch_read error to aos::Error

File: src/communication/vchannel.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    if (static_cast<size_t>(ret) != size) {
+        err = data.Resize(ret);
+        if (!err.IsNone()) {
+            return AOS_ERROR_WRAP(err);
+        }
+
+        return AOS_ERROR_WRAP(aos::ErrorEnum::eFailed);
```
Comment (mykola-kobets-epam) [inline src/communication/vchannel.cpp:65]: not sure we should do resize if anyway return Failed.

File: src/communication/vchannel.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error VChannel::Write(const aos::Array<uint8_t>& data)
+{
+    auto ret = vch_write(&mWriteHandle, data.Get(), data.Size());
+    if (ret < 0) {
+        return AOS_ERROR_WRAP(ret);
```
Comment (mykola-kobets-epam) [inline src/communication/vchannel.cpp:75]: there could be a convertion problem I think.

Comment (al1img) [reply src/communication/vchannel.cpp:75]: There is no. ret is errno that is handled by aos::Error with the dedicated constructor.

File: src/communication/communication.cpp [hunk 14]
```diff
@@ -161,18 +168,96 @@ void Communication::ChannelHandler(Channel channel)
...
+        static_cast<int32_t>(messageError.Value())};
+
+    auto checksum = CalculateSha256(data);
+    if (!checksum.mError.IsNone()) {
+        return AOS_ERROR_WRAP(checksum.mError);
+    }
+
+    aos::Array<uint8_t>(reinterpret_cast<uint8_t*>(header.mSha256), aos::cSHA256Size) = checksum.mValue;
+
+    auto err = mChannels[channel]->Write(
```
[the remark this answers is not in the document]
Comment (al1img) [reply src/communication/communication.cpp:222]: I don't think we should perform such a check for interface functions. It is not supposed this function is called by external modules/classes etc.

File: .vscode/settings.json (hunk 1 above)
Comment (al1img) [reply .vscode/settings.json:7]: committed by mistake - removed.

File: src/domains/domd/domain_bins.S (hunk 4 above)
Comment (al1img) [reply src/domains/domd/domain_bins.S:23]: mistake

File: src/clocksync/clocksync.hpp (hunk 3 above)
Comment (al1img) [reply src/clocksync/clocksync.hpp:67]: right. removed.

File: tests/communication/CMakeLists.txt (hunk 6 above)
Comment (al1img) [reply tests/communication/CMakeLists.txt]: Which library do you mean?

File: .vscode/c_cpp_properties.json (hunk 2 above)
Comment (al1img) [reply .vscode/c_cpp_properties.json]: done

File: src/communication/vchannel.cpp (hunk 12 above)
Comment (al1img) [reply src/communication/vchannel.cpp:65]: The idea is: if not all requested data is read, return what is read + error. How this data will be used is up to the client.

File: src/utils/pbconvert.hpp (hunk 8 above)
Comment (al1img) [reply src/utils/pbconvert.hpp]: Yes, but I don't want to check error for pbconvert function, otherwise it will bloat the code drastically. Assert added.

File: src/utils/pbconvert.hpp (hunk 7 above)
Comment (al1img) [reply src/utils/pbconvert.hpp:40]: Two points:
* if assert occurs - system is not usable any more;
* assert is enabled in debug mode only. For production, all assert will be disabled.

So I will add assert additionally to reducing size.

File: src/communication/vchannel.cpp (hunk 11 above)
Comment (al1img) [reply src/communication/vchannel.cpp:56]: there is.

File: src/utils/pbconvert.hpp (hunk 9 above)
Comment (al1img) [reply src/utils/pbconvert.hpp:90]: Why it is strange? PB declares different types for byte arrays with different size. I have no other options.

**Review by MykolaSolyanko — APPROVED** [review]

File: tests/communication/CMakeLists.txt (hunk 6 above)
Comment (mykola-kobets-epam) [reply tests/communication/CMakeLists.txt]: For current project I would suggest creating "communication" static library & link it to app & to the test. In this case it wouldn't be necessary to duplicate required source files & compiler options

File: src/utils/pbconvert.hpp (hunk 9 above)
Comment (mykola-kobets-epam) [reply src/utils/pbconvert.hpp:90]: understood

File: src/communication/vchannel.cpp (hunk 12 above)
Comment (mykola-kobets-epam) [reply src/communication/vchannel.cpp:65]: got it.

File: src/communication/vchannel.cpp (hunk 13 above)
Comment (mykola-kobets-epam) [reply src/communication/vchannel.cpp:75]: ok

File: src/communication/vchannel.cpp (hunk 11 above)
Comment (mykola-kobets-epam) [reply src/communication/vchannel.cpp:56]: ok

File: src/utils/pbconvert.hpp (hunk 7 above)
Comment (mykola-kobets-epam) [reply src/utils/pbconvert.hpp:40]: ok

File: src/resourcemanager/resourcemanager.hpp (hunk 5 above)
Comment (mykola-kobets-epam) [reply src/resourcemanager/resourcemanager.hpp:17]: ok

**Review by mykola-kobets-epam — APPROVED** [review]
